File: minijinja/src/defaults.rs
use std::borrow::Cow;
use std::collections::BTreeMap;

use crate::error::Error;
use crate::filters::{self, BoxedFilter};
use crate::output::Output;
use crate::tests::{self, BoxedTest};
use crate::utils::{write_escaped, AutoEscape};
use crate::value::Value;
use crate::vm::State;

pub(crate) fn no_auto_escape(_: &str) -> AutoEscape {
    AutoEscape::None
}

/// The default logic for auto escaping based on file extension.
///
/// * [`Html`](AutoEscape::Html): `.html`, `.htm`, `.xml`
#[cfg_attr(
    feature = "json",
    doc = r" * [`Json`](AutoEscape::Json): `.json`, `.json5`, `.js`, `.yaml`, `.yml`"
)]
/// * [`None`](AutoEscape::None): _all others_
///
/// Additionally `.j2` as final extension is ignored. So `.html.j2` is the
/// considered the same as `.html`.
pub fn default_auto_escape_callback(name: &str) -> AutoEscape {
    match name.strip_suffix(".j2").unwrap_or(name).rsplit('.').next() {
        Some("html" | "htm" | "xml") => AutoEscape::Html,
        #[cfg(feature = "json")]
        Some("json" | "json5" | "js" | "yaml" | "yml") => AutoEscape::Json,
        _ => AutoEscape::None,
    }
}

/// The default formatter.
///
/// This formatter takes a value and directly writes it into the output format
/// while honoring the requested auto escape format of the state.  If the
/// value is already marked as safe, it's handled as if no auto escaping
/// was requested.
///
/// * [`Html`](AutoEscape::Html): performs HTML escaping
#[cfg_attr(
    feature = "json",
    doc = r" * [`Json`](AutoEscape::Json): serializes values to JSON"
)]
/// * [`None`](AutoEscape::None): no escaping
/// * [`Custom(..)`](AutoEscape::Custom): results in an error
pub fn escape_formatter(out: &mut Output, state: &State, value: &Value) -> Result<(), Error> {
    write_escaped(out, state.auto_escape(), value)
}

pub(crate) fn get_builtin_filters() -> BTreeMap<Cow<'static, str>, filters::BoxedFilter> {
    let mut rv = BTreeMap::new();
    rv.insert("safe".into(), BoxedFilter::new(filters::safe));
    let escape = BoxedFilter::new(filters::escape);
    rv.insert("escape".into(), escape.clone());
    rv.insert("e".into(), escape);
    #[cfg(feature = "builtins")]
    {
        rv.insert("lower".into(), BoxedFilter::new(filters::lower));
        rv.insert("upper".into(), BoxedFilter::new(filters::upper));
        rv.insert("title".into(), BoxedFilter::new(filters::title));
        rv.insert("capitalize".into(), BoxedFilter::new(filters::capitalize));
        rv.insert("replace".into(), BoxedFilter::new(filters::replace));
        let length = BoxedFilter::new(filters::length);
        rv.insert("length".into(), length.clone());
        rv.insert("count".into(), length);
        rv.insert("dictsort".into(), BoxedFilter::new(filters::dictsort));
        rv.insert("items".into(), BoxedFilter::new(filters::items));
        rv.insert("reverse".into(), BoxedFilter::new(filters::reverse));
        rv.insert("trim".into(), BoxedFilter::new(filters::trim));
        rv.insert("join".into(), BoxedFilter::new(filters::join));
        rv.insert("split".into(), BoxedFilter::new(filters::split));
        rv.insert("lines".into(), BoxedFilter::new(filters::lines));
        rv.insert("default".into(), BoxedFilter::new(filters::default));
        rv.insert("round".into(), BoxedFilter::new(filters::round));
        rv.insert("abs".into(), BoxedFilter::new(filters::abs));
        rv.insert("int".into(), BoxedFilter::new(filters::int));
        rv.insert("float".into(), BoxedFilter::new(filters::float));
        rv.insert("attr".into(), BoxedFilter::new(filters::attr));
        rv.insert("first".into(), BoxedFilter::new(filters::first));
        rv.insert("last".into(), BoxedFilter::new(filters::last));
        rv.insert("min".into(), BoxedFilter::new(filters::min));
        rv.insert("max".into(), BoxedFilter::new(filters::max));
        rv.insert("sort".into(), BoxedFilter::new(filters::sort));
        rv.insert("d".into(), BoxedFilter::new(filters::default));
        rv.insert("list".into(), BoxedFilter::new(filters::list));
        rv.insert("string".into(), BoxedFilter::new(filters::string));
        rv.insert("bool".into(), BoxedFilter::new(filters::bool));
        rv.insert("batch".into(), BoxedFilter::new(filters::batch));
        rv.insert("slice".into(), BoxedFilter::new(filters::slice));
        rv.insert("indent".into(), BoxedFilter::new(filters::indent));
        rv.insert("select".into(), BoxedFilter::new(filters::select));
        rv.insert("reject".into(), BoxedFilter::new(filters::reject));
        rv.insert("selectattr".into(), BoxedFilter::new(filters::selectattr));
        rv.insert("rejectattr".into(), BoxedFilter::new(filters::rejectattr));
        rv.insert("map".into(), BoxedFilter::new(filters::map));
        rv.insert("groupby".into(), BoxedFilter::new(filters::groupby));
        rv.insert("unique".into(), BoxedFilter::new(filters::unique));
        rv.insert("pprint".into(), BoxedFilter::new(filters::pprint));

        #[cfg(feature = "json")]
        {
            rv.insert("tojson".into(), BoxedFilter::new(filters::tojson));
        }
        #[cfg(feature = "urlencode")]
        {
            rv.insert("urlencode".into(), BoxedFilter::new(filters::urlencode));
        }
    }

    rv
}

pub(crate) fn get_builtin_tests() -> BTreeMap<Cow<'static, str>, BoxedTest> {
    let mut rv = BTreeMap::new();
    rv.insert("undefined".into(), BoxedTest::new(tests::is_undefined));
    rv.insert("defined".into(), BoxedTest::new(tests::is_defined));
    rv.insert("none".into(), BoxedTest::new(tests::is_none));
    let is_safe = BoxedTest::new(tests::is_safe);
    rv.insert("safe".into(), is_safe.clone());
    rv.insert("escaped".into(), is_safe);
    #[cfg(feature = "builtins")]
    {
        rv.insert("boolean".into(), BoxedTest::new(tests::is_boolean));
        rv.insert("odd".into(), BoxedTest::new(tests::is_odd));
        rv.insert("even".into(), BoxedTest::new(tests::is_even));
        rv.insert("divisibleby".into(), BoxedTest::new(tests::is_divisibleby));
        rv.insert("number".into(), BoxedTest::new(tests::is_number));
        rv.insert("integer".into(), BoxedTest::new(tests::is_integer));
        rv.insert("int".into(), BoxedTest::new(tests::is_integer));
        rv.insert("float".into(), BoxedTest::new(tests::is_float));
        rv.insert("string".into(), BoxedTest::new(tests::is_string));
        rv.insert("sequence".into(), BoxedTest::new(tests::is_sequence));
        rv.insert("iterable".into(), BoxedTest::new(tests::is_iterable));
        rv.insert("mapping".into(), BoxedTest::new(tests::is_mapping));
        rv.insert(
            "startingwith".into(),
            BoxedTest::new(tests::is_startingwith),
        );
        rv.insert("endingwith".into(), BoxedTest::new(tests::is_endingwith));
        rv.insert("lower".into(), BoxedTest::new(tests::is_lower));
        rv.insert("upper".into(), BoxedTest::new(tests::is_upper));
        rv.insert("sameas".into(), BoxedTest::new(tests::is_sameas));

        // operators
        let is_eq = BoxedTest::new(tests::is_eq);
        rv.insert("eq".into(), is_eq.clone());
        rv.insert("equalto".into(), is_eq.clone());
        rv.insert("==".into(), is_eq);
        let is_ne = BoxedTest::new(tests::is_ne);
        rv.insert("ne".into(), is_ne.clone());
        rv.insert("!=".into(), is_ne);
        let is_lt = BoxedTest::new(tests::is_lt);
        rv.insert("lt".into(), is_lt.clone());
        rv.insert("lessthan".into(), is_lt.clone());
        rv.insert("<".into(), is_lt);
        let is_le = BoxedTest::new(tests::is_le);
        rv.insert("le".into(), is_le.clone());
        rv.insert("<=".into(), is_le);
        let is_gt = BoxedTest::new(tests::is_gt);
        rv.insert("gt".into(), is_gt.clone());
        rv.insert("greaterthan".into(), is_gt.clone());
        rv.insert(">".into(), is_gt);
        let is_ge = BoxedTest::new(tests::is_ge);
        rv.insert("ge".into(), is_ge.clone());
        rv.insert(">=".into(), is_ge);
        rv.insert("in".into(), BoxedTest::new(tests::is_in));
        rv.insert("true".into(), BoxedTest::new(tests::is_true));
        rv.insert("false".into(), BoxedTest::new(tests::is_false));
        rv.insert("filter".into(), BoxedTest::new(tests::is_filter));
        rv.insert("test".into(), BoxedTest::new(tests::is_test));
    }
    rv
}

pub(crate) fn get_globals() -> BTreeMap<Cow<'static, str>, Value> {
    #[allow(unused_mut)]
    let mut rv = BTreeMap::new();
    #[cfg(feature = "builtins")]
    {
        use crate::functions::{self, BoxedFunction};
        rv.insert(
            "range".into(),
            BoxedFunction::new(functions::range).to_value(),
        );
        rv.insert(
            "dict".into(),
            BoxedFunction::new(functions::dict).to_value(),
        );
        rv.insert(
            "debug".into(),
            BoxedFunction::new(functions::debug).to_value(),
        );
        rv.insert(
            "namespace".into(),
            BoxedFunction::new(functions::namespace).to_value(),
        );
    }

    rv
}

#[cfg(test)]
mod unit_tests {
    use super::*;

    use similar_asserts::assert_eq;

    #[test]
    fn test_default_uto_escape() {
        assert_eq!(default_auto_escape_callback("foo.html"), AutoEscape::Html);
        assert_eq!(
            default_auto_escape_callback("foo.html.j2"),
            AutoEscape::Html
        );
        assert_eq!(default_auto_escape_callback("foo.htm"), AutoEscape::Html);
        assert_eq!(default_auto_escape_callback("foo.htm.j2"), AutoEscape::Html);
        assert_eq!(default_auto_escape_callback("foo.txt"), AutoEscape::None);
        assert_eq!(default_auto_escape_callback("foo.txt.j2"), AutoEscape::None);

        #[cfg(feature = "json")]
        {
            assert_eq!(default_auto_escape_callback("foo.yaml"), AutoEscape::Json);
            assert_eq!(
                default_auto_escape_callback("foo.json.j2"),
                AutoEscape::Json
            );
        }
    }
}

File: minijinja-go/defaults.go
package minijinja

import (
	"fmt"
	"sort"
	"strings"

	"github.com/mitsuhiko/minijinja/minijinja-go/v2/filters"
	"github.com/mitsuhiko/minijinja/minijinja-go/v2/tests"
	"github.com/mitsuhiko/minijinja/minijinja-go/v2/value"
)

func registerDefaultFilters(env *Environment) {
	// String filters
	env.AddFilter("upper", filters.FilterUpper)
	env.AddFilter("lower", filters.FilterLower)
	env.AddFilter("capitalize", filters.FilterCapitalize)
	env.AddFilter("title", filters.FilterTitle)
	env.AddFilter("trim", filters.FilterTrim)
	env.AddFilter("replace", filters.FilterReplace)
	env.AddFilter("format", filters.FilterFormat)
	env.AddFilter("default", filters.FilterDefault)
	env.AddFilter("d", filters.FilterDefault) // alias
	env.AddFilter("safe", filters.FilterSafe)
	env.AddFilter("escape", filters.FilterEscape)
	env.AddFilter("e", filters.FilterEscape) // alias
	env.AddFilter("string", filters.FilterString)
	env.AddFilter("bool", filters.FilterBool)
	env.AddFilter("split", filters.FilterSplit)
	env.AddFilter("lines", filters.FilterLines)

	// List/sequence filters
	env.AddFilter("length", filters.FilterLength)
	env.AddFilter("count", filters.FilterLength) // alias
	env.AddFilter("first", filters.FilterFirst)
	env.AddFilter("last", filters.FilterLast)
	env.AddFilter("reverse", filters.FilterReverse)
	env.AddFilter("sort", filters.FilterSort)
	env.AddFilter("join", filters.FilterJoin)
	env.AddFilter("list", filters.FilterList)
	env.AddFilter("unique", filters.FilterUnique)
	env.AddFilter("min", filters.FilterMin)
	env.AddFilter("max", filters.FilterMax)
	env.AddFilter("sum", filters.FilterSum)
	env.AddFilter("batch", filters.FilterBatch)
	env.AddFilter("slice", filters.FilterSlice)
	env.AddFilter("map", filters.FilterMap)
	env.AddFilter("select", filters.FilterSelect)
	env.AddFilter("reject", filters.FilterReject)
	env.AddFilter("selectattr", filters.FilterSelectAttr)
	env.AddFilter("rejectattr", filters.FilterRejectAttr)
	env.AddFilter("groupby", filters.FilterGroupBy)
	env.AddFilter("chain", filters.FilterChain)
	env.AddFilter("zip", filters.FilterZip)

	// Numeric filters
	env.AddFilter("abs", filters.FilterAbs)
	env.AddFilter("int", filters.FilterInt)
	env.AddFilter("float", filters.FilterFloat)
	env.AddFilter("round", filters.FilterRound)

	// Dict filters
	env.AddFilter("items", filters.FilterItems)
	env.AddFilter("dictsort", filters.FilterDictSort)

	// Other filters
	env.AddFilter("attr", filters.FilterAttr)
	env.AddFilter("indent", filters.FilterIndent)
	env.AddFilter("pprint", filters.FilterPprint)

	// JSON and URL filters
	env.AddFilter("tojson", filters.FilterTojson)
	env.AddFilter("urlencode", filters.FilterUrlencode)
}

func registerDefaultTests(env *Environment) {
	env.AddTest("defined", tests.TestDefined)
	env.AddTest("undefined", tests.TestUndefined)
	env.AddTest("none", tests.TestNone)
	env.AddTest("true", tests.TestTrue)
	env.AddTest("false", tests.TestFalse)
	env.AddTest("odd", tests.TestOdd)
	env.AddTest("even", tests.TestEven)
	env.AddTest("divisibleby", tests.TestDivisibleBy)
	env.AddTest("eq", tests.TestEq)
	env.AddTest("equalto", tests.TestEq)
	env.AddTest("==", tests.TestEq)
	env.AddTest("ne", tests.TestNe)
	env.AddTest("!=", tests.TestNe)
	env.AddTest("lt", tests.TestLt)
	env.AddTest("lessthan", tests.TestLt)
	env.AddTest("<", tests.TestLt)
	env.AddTest("le", tests.TestLe)
	env.AddTest("<=", tests.TestLe)
	env.AddTest("gt", tests.TestGt)
	env.AddTest("greaterthan", tests.TestGt)
	env.AddTest(">", tests.TestGt)
	env.AddTest("ge", tests.TestGe)
	env.AddTest(">=", tests.TestGe)
	env.AddTest("in", tests.TestIn)
	env.AddTest("string", tests.TestString)
	env.AddTest("number", tests.TestNumber)
	env.AddTest("integer", tests.TestInteger)
	env.AddTest("int", tests.TestInteger) // alias
	env.AddTest("float", tests.TestFloat)
	env.AddTest("boolean", tests.TestBoolean)
	env.AddTest("sequence", tests.TestSequence)
	env.AddTest("mapping", tests.TestMapping)
	env.AddTest("iterable", tests.TestIterable)
	env.AddTest("startingwith", tests.TestStartingWith)
	env.AddTest("endingwith", tests.TestEndingWith)
	env.AddTest("containing", tests.TestContaining)
	env.AddTest("safe", tests.TestSafe)
	env.AddTest("escaped", tests.TestSafe) // alias
	env.AddTest("sameas", tests.TestSameAs)
	env.AddTest("lower", tests.TestLower)
	env.AddTest("upper", tests.TestUpper)
	env.AddTest("filter", tests.TestFilter)
	env.AddTest("test", tests.TestTest)
}

func registerDefaultFunctions(env *Environment) {
	env.AddFunction("range", fnRange)
	env.AddFunction("dict", fnDict)
	env.AddFunction("cycler", fnCycler)
	env.AddFunction("joiner", fnJoiner)
	env.AddFunction("namespace", fnNamespace)
	env.AddFunction("debug", fnDebug)
	env.AddFunction("lipsum", fnLipsum)
}

// --- Functions ---

func fnRange(_ *State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	var start, stop, step int64 = 0, 0, 1

	switch len(args) {
	case 1:
		stop, _ = args[0].AsInt()
	case 2:
		start, _ = args[0].AsInt()
		stop, _ = args[1].AsInt()
	case 3:
		start, _ = args[0].AsInt()
		stop, _ = args[1].AsInt()
		step, _ = args[2].AsInt()
	default:
		return value.FromIterator(value.NewIterator("range", nil)), nil
	}

	if step == 0 {
		return value.Undefined(), fmt.Errorf("range step cannot be zero")
	}

	length := int64(0)
	if step > 0 {
		if stop > start {
			length = (stop - start + step - 1) / step
		}
	} else {
		if stop < start {
			negStep := -step
			length = (start - stop + negStep - 1) / negStep
		}
	}
	if length > 100000 {
		return value.Undefined(), NewError(ErrInvalidOperation, "range has too many elements")
	}

	var result []value.Value
	if step > 0 {
		for i := start; i < stop; i += step {
			result = append(result, value.FromInt(i))
		}
	} else {
		for i := start; i > stop; i += step {
			result = append(result, value.FromInt(i))
		}
	}
	return value.FromIterator(value.NewIterator("range", result)), nil
}

func fnDict(_ *State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	result := make(map[string]value.Value)

	// First, copy from first positional argument if it's a map
	if len(args) > 0 {
		if m, ok := args[0].AsMap(); ok {
			for k, v := range m {
				result[k] = v
			}
		} else {
			// Try to iterate as items
			items := args[0].Iter()
			if items != nil {
				for _, item := range items {
					if pair, ok := item.AsSlice(); ok && len(pair) == 2 {
						if k, ok := pair[0].AsString(); ok {
							result[k] = pair[1]
						}
					}
				}
			}
		}
	}

	// Then apply kwargs (overwriting)
	for k, v := range kwargs {
		result[k] = v
	}
	return value.FromMap(result), nil
}

func fnCycler(_ *State, args []value.Value, _ map[string]value.Value) (value.Value, error) {
	return value.FromObject(&cyclerObject{
		items: args,
		index: 0,
	}), nil
}

// cyclerObject implements a cycler that cycles through values
type cyclerObject struct {
	items []value.Value
	index int
}

func (c *cyclerObject) GetAttr(name string) value.Value {
	switch name {
	case "next":
		return value.FromCallable(&cyclerNextCallable{cycler: c})
	case "current":
		if len(c.items) == 0 {
			return value.Undefined()
		}
		idx := c.index
		if idx == 0 {
			idx = len(c.items)
		}
		return c.items[idx-1]
	case "reset":
		return value.FromCallable(&cyclerResetCallable{cycler: c})
	}
	return value.Undefined()
}

type cyclerNextCallable struct {
	cycler *cyclerObject
}

func (c *cyclerNextCallable) Call(state value.State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	_ = state
	if len(c.cycler.items) == 0 {
		return value.Undefined(), nil
	}
	result := c.cycler.items[c.cycler.index]
	c.cycler.index = (c.cycler.index + 1) % len(c.cycler.items)
	return result, nil
}

type cyclerResetCallable struct {
	cycler *cyclerObject
}

func (c *cyclerResetCallable) Call(state value.State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	_ = state
	c.cycler.index = 0
	return value.Undefined(), nil
}

func fnJoiner(_ *State, args []value.Value, _ map[string]value.Value) (value.Value, error) {
	sep := ", "
	if len(args) > 0 {
		if s, ok := args[0].AsString(); ok {
			sep = s
		}
	}
	return value.FromCallable(&joinerCallable{
		sep:   sep,
		first: true,
	}), nil
}

// joinerCallable implements a joiner that returns separator after first call
type joinerCallable struct {
	sep   string
	first bool
}

func (j *joinerCallable) Call(state value.State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	_ = state
	if j.first {
		j.first = false
		return value.FromString(""), nil
	}
	return value.FromString(j.sep), nil
}

func fnNamespace(_ *State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	ns := &namespaceValue{
		data: make(map[string]value.Value),
	}

	// If first argument is a map, copy from it
	if len(args) > 0 {
		if m, ok := args[0].AsMap(); ok {
			for k, v := range m {
				ns.data[k] = v
			}
		} else if !args[0].IsUndefined() && !args[0].IsNone() {
			return value.Undefined(), NewError(ErrInvalidOperation, "namespace expects a mapping")
		}
	}

	// Apply kwargs
	for k, v := range kwargs {
		ns.data[k] = v
	}
	return value.FromObject(ns), nil
}

// namespaceValue is a mutable namespace object
type namespaceValue struct {
	data map[string]value.Value
}

func (n *namespaceValue) GetAttr(name string) value.Value {
	if v, ok := n.data[name]; ok {
		return v
	}
	return value.Undefined()
}

func (n *namespaceValue) SetAttr(name string, val value.Value) {
	n.data[name] = val
}

func (n *namespaceValue) String() string {
	keys := make([]string, 0, len(n.data))
	for k := range n.data {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, fmt.Sprintf("%q: %s", k, n.data[k].Repr()))
	}
	return "{" + strings.Join(parts, ", ") + "}"
}

func (n *namespaceValue) Map() map[string]value.Value {
	return n.data
}

func fnDebug(state *State, args []value.Value, _ map[string]value.Value) (value.Value, error) {
	// If arguments provided, debug those values
	if len(args) > 0 {
		var parts []string
		for _, arg := range args {
			parts = append(parts, arg.Repr())
		}
		return value.FromString(strings.Join(parts, ", ")), nil
	}

	// Otherwise debug the current state
	var parts []string
	parts = append(parts, fmt.Sprintf("State {"))
	parts = append(parts, fmt.Sprintf("  name: %q,", state.name))
	parts = append(parts, "  current variables: {")

	// Collect variables from scopes
	for i := len(state.scopes) - 1; i >= 0; i-- {
		keys := make([]string, 0, len(state.scopes[i]))
		for k := range state.scopes[i] {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			parts = append(parts, fmt.Sprintf("    %s: %s,", k, state.scopes[i][k].Repr()))
		}
	}
	if !state.rootContext.IsUndefined() {
		if obj, ok := state.rootContext.AsObject(); ok {
			if m, ok := obj.(value.MapObject); ok {
				keys := m.Keys()
				sort.Strings(keys)
				for _, k := range keys {
					parts = append(parts, fmt.Sprintf("    %s: %s,", k, state.rootContext.GetAttr(k).Repr()))
				}
			}
		}
	}
	parts = append(parts, "  }")
	parts = append(parts, "}")

	return value.FromString(strings.Join(parts, "\n")), nil
}

func fnLipsum(_ *State, args []value.Value, kwargs map[string]value.Value) (value.Value, error) {
	n := int64(5)
	if len(args) > 0 {
		if nn, ok := args[0].AsInt(); ok {
			n = nn
		}
	}
	if nn, ok := kwargs["n"]; ok {
		if nnn, ok := nn.AsInt(); ok {
			n = nnn
		}
	}

	lorem := "Lorem ipsum dolor sit amet, consectetur adipiscing elit. " +
		"Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. " +
		"Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris. "

	var result strings.Builder
	for i := int64(0); i < n; i++ {
		if i > 0 {
			result.WriteString("\n\n")
		}
		result.WriteString(lorem)
	}
	return value.FromSafeString("<p>" + result.String() + "</p>"), nil
}
